### PR TITLE
Add error handling

### PR DESCRIPTION
**Pull request by calippo**
- First commit is the logic
- Secondo commit is just refactoring of modules

# test plan

tested locally on `testo` project

### CODE REVIEW

**Comment by calippo** [comment]
Endpoints

```scala
package testo
import tapir._
import tapir.Codec.JsonCodec

class ControllerEndpoints(statusCodes: String => Int = _ => 422)(
    implicit spittyCash: JsonCodec[SpittyCash],
    errorone: JsonCodec[Errorone],
    errorino: JsonCodec[Errorino],
    string: JsonCodec[String]
) {

  val ghetto: Endpoint[(Int, String), Errore, SpittyCash, Nothing] =
    endpoint.get
      .in("ghetto")
      .in(query[Int]("i"))
      .in(query[String]("s"))
      .errorOut(
        oneOf[Errore](
          statusMapping(statusCodes("Errorone"), jsonBody[Errorone]),
          statusMapping(statusCodes("Errorino"), jsonBody[Errorino])
        )
      )
      .out(jsonBody[SpittyCash])

  val pusho: Endpoint[SpittyCash, String, String, Nothing] = endpoint.post
    .in("pusho")
    .in(jsonBody[SpittyCash])
    .errorOut(stringBody)
    .out(jsonBody[String])
}
```

Http4s
```scala
package testo
import cats.effect._
import cats.implicits._
import cats.data.NonEmptyList
import org.http4s._
import org.http4s.implicits._
import tapir.server.http4s._
import tapir.Codec.JsonCodec

class ControllerHttp4sEndpoints(controller: Controller[IO])(
    implicit spittyCash: JsonCodec[SpittyCash],
    errorone: JsonCodec[Errorone],
    errorino: JsonCodec[Errorino],
    string: JsonCodec[String],
    io: ContextShift[IO]
) {
  private[this] val endpoints = new ControllerEndpoints()
  private[this] val ghetto: HttpRoutes[IO] =
    endpoints.ghetto.toRoutes((controller.ghetto _).tupled)
  private[this] val pusho: HttpRoutes[IO] =
    endpoints.pusho.toRoutes(controller.pusho)
  val app: HttpApp[IO] = NonEmptyList(ghetto, List(pusho)).reduceK.orNotFound
}

```

File: tapiro/core/src/main/scala/io/buildo/tapiro/Http4sMeta.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  val `class` = (
+    `package`: Term.Name,
+    controllerName: Type.Name,
+    endpointsName: Type.Name,
+    implicits: List[Term.Param],
+    http4sEndpoints: List[Defn.Val],
+    app: Defn.Val,
+  ) => {
+    val tapirEndpoints = q"private[this] val endpoints = new $endpointsName()"
+    val httpsEndpointsName = Type.Name(s"${controllerName.syntax}Http4sEndpoints")
```
Comment (gabro) [inline tapiro/core/src/main/scala/io/buildo/tapiro/Http4sMeta.scala]:
```suggestion
    val httpsEndpointsName = t"${controllerName.syntax}Http4sEndpoints"
```

File: tapiro/core/src/main/scala/io/buildo/tapiro/Meta.scala [hunk 2]
```diff
@@ -1,170 +1,32 @@
...
+  private[this] val typeToImplicitParam = (tpe: MetarpheusType) =>
+    stringToImplicitParam(typeNameString(tpe))
 
-  val routeToTapirEndpoint = (route: Route) =>
-    q"val ${Pat.Var(Term.Name(route.name.tail.mkString))}: ${endpointType(route)} = ${endpointImpl(route)}"
-
-  private[this] val typeToImplicitParam = (tpe: MetarpheusType) => {
-    val name = typeNameString(tpe)
+  private[this] val stringToImplicitParam = (name: String) => {
     val paramName = Term.Name(s"${name.head.toLower}${name.tail}")
```
Comment (gabro) [inline tapiro/core/src/main/scala/io/buildo/tapiro/Meta.scala]:
```suggestion
     val paramName = q"${name.head.toLower}${name.tail}"
```

File: tapiro/core/src/main/scala/io/buildo/tapiro/TapirMeta.scala [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+  class $name(statusCodes: String => Int = _ => 422)(..$implicits) {
+   ..$body
+  }
+}
+""",
+  }
+
+  val routeToTapirEndpoint = (route: TapiroRoute) =>
+    q"val ${Pat.Var(Term.Name(route.route.name.tail.mkString))}: ${endpointType(route.route)} = ${endpointImpl(route)}"
```
Comment (gabro) [inline tapiro/core/src/main/scala/io/buildo/tapiro/TapirMeta.scala]:
```suggestion
    val valName = p"route.route.name.tail.mkString"
    q"val $valName: ${endpointType(route.route)} = ${endpointImpl(route)}"
```

File: tapiro/core/src/main/scala/io/buildo/tapiro/TapirMeta.scala [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    val error = Type.Name(route.error.map(typeNameString).getOrElse("String"))
+    t"Endpoint[$argsType, $error, $returnType, Nothing]"
+  }
+
+  private[this] val endpointImpl = (route: TapiroRoute) => {
+    val basicEndpoint = Term.Apply(
+      Term.Select(Term.Select(Term.Name("endpoint"), Term.Name(route.route.method)), Term.Name("in")),
+      List(Lit.String(route.route.name.tail.mkString)),
+    )
```
Comment (gabro) [inline tapiro/core/src/main/scala/io/buildo/tapiro/TapirMeta.scala]: This probably works:

```scala
q"endpoint.${route.route.method}.in(${route.route.name.tail.mkString})"
```

**Comment by calippo** [comment]
@gabro should we merge this?

**Comment by gabro** [comment]
Yep

**Pull request merged**